src/components/progress-report/workflow/handlers/progress-report-workflow-notification.handler.ts

```diff
@@ -0,0 +1,231 @@
+    const usersInEnvList: Recipient[] = await Promise.all(
+      envEmailList.map(async (item) => {
+        const user = await this.repo.getUserInfoByEmail(item);
+        if (user.length > 0) {
+          const [{ id, email }] = user;
+          return { id: id, email: email };
+        } else {
+          return { id: undefined, email: item };
+        }
+      }),
+    );
+
+    const recipients = uniqBy(
+      [...usersInEnvList, ...membersWithValidRoles],
+      'email',
+    );
+    const notifications: EmailReportStatusNotification[] = await Promise.all(
+      recipients.map((recipient) =>
+        this.prepareNotificationObject(
+          session,
+          reportId,
+          previousStatus,
+          recipient,
+          workflowEvent,
+        ),
+      ),
+    );
+
+    this.logger.info('Notifying', {
+      emails: notifications.map((r) => r.recipient.email.value),
+      projectId: notifications[0]?.project.id ?? undefined,
+      languageId: notifications[0]?.language.id ?? undefined,
+      reportId: reportId,
+      newStatusVal: notifications[0]?.newStatusVal ?? undefined,
+      previousStatusVal: notifications[0]?.previousStatusVal ?? undefined,
+    });
+
+    for (const notification of notifications) {
+      if (notification.recipient.email.value) {
+        await this.emailService.send(
+          notification.recipient.email.value,
+          ProgressReportStatusChanged,
+          notification,
+        );
+      }
+    }
+  }
+
+  private async prepareNotificationObject(
+    session: Session,
+    reportId: ID,
+    previousStatus: Status,
+    receiver: Recipient,
+    unsecuredEvent: UnsecuredDto<ProgressReportWorkflowEvent>,
+  ): Promise<EmailReportStatusNotification> {
+    const { projectId, languageId } = await this.repo.getProjectInfoByReportId(
+      reportId,
+    );
+
+    const recipientId = receiver.id
+      ? receiver.id
+      : this.configService.rootAdmin.id;
+    const recipientSession = await this.auth.sessionForUser(recipientId);
+
+    const recipient = receiver.id
+      ? await this.userService.readOne(recipientId, recipientSession)
+      : this.mapRecipient(receiver);
+
+    const project = await this.projectService.readOne(
+      projectId,
+      recipientSession,
+    );
+    const language = await this.languageService.readOne(
+      languageId,
+      recipientSession,
+    );
+    const report = await this.reportService.readOne(reportId, recipientSession);
+    const changedBy = await this.userService.readOne(
+      session.userId,
+      recipientSession,
+    );
+    const workflowEvent = this.workflowService.secure(
+      unsecuredEvent,
+      recipientSession,
+    );
+
+    return {
+      changedBy,
+      recipient,
+      project,
+      language,
+      report,
+      newStatusVal: report.status?.value,
+      previousStatusVal: previousStatus,
+      workflowEvent: workflowEvent,
+    };
+  }
+
+  private mapRecipient(receiver: Recipient) {
```

Comment:
This is kinda a meaningless name and doesn't apply to every recipient. Explicit return type here will help make the connection to where it's going and help your IDE help you and provide better type errors.
```suggestion
  private fakeUserFromEmailAddress(
    receiver: Recipient,
  ): EmailReportStatusNotification['recipient'] {
```